Make (most) UI elements reachable via keyboard

Leimi opened the pull request:
Hello,

Here is a pull request to globally improve keyboard navigation:

- the PR improves navigation when popovers, drawers, and dialogs open and hide. For example, now when a Dialog closes, it tries to put back focus on the button that triggered it. I feel we could talk about it more because 1) there is still room for improvement, and 2) maybe my ideas to fix them are debatable :) For example I'm not entirely satisfied with my Popover focus trap solution. I will elaborate more on that soon,
- the PR tries to make every interactive element in Jitsi Meet reachable via keyboard. I'm sure I missed some stuff as I don't fully know the app, but there is already lots of stuff that are improved.

The list of improved stuff for keyboard users:

- Popovers are focused directly on open (instead of needing to tab all the way through the bottom of the DOM) and focus is put back on the element that opened them on close
- items in the toolbox's overflow menu are now reachable via keyboard
- Dialogs try to focus back the element that opened them on close
- Drawers try to focus back the element that opened them on close
- it's now possible to pin/unpin participants in the filmstrip thumbnails, via keyboard
- the connection info and menu in thumbnails that only appeared on mouse hover before, are now reachable via keyboard
- in the conference info box, the performance settings button and the few other interactive labels are now actionable via keyboard
- in the polls pane, the details/results buttons are now reachable via keyboard
- in the participants pane, the actions are now reachable via keyboard 


I know there are lots of commits already but they are well split and described in their messages, feeling it will be okay to review commits one by one :pray: 

Note this PR only handle the actual reaching of things with keyboard. It doesn't impact at all the text that is announced by screen readers when we are focusing things. This will need another PR.

This is related to the PR I closed earlier: https://github.com/jitsi/jitsi-meet/pull/12644 and the issue on accessibility: https://github.com/jitsi/jitsi-meet/issues/12379

ping @saghul

jitsi-jenkins commented:
Hi, thanks for your contribution!
If you haven't already done so, could you please make sure you sign our CLA (https://jitsi.org/icla for individuals and https://jitsi.org/ccla for corporations)? We would unfortunately be unable to merge your patch unless we have that piece :(.

Calinteodor commented:
@Leimi Reviewed this one, too. 
Thanks again and have a nice weekend!

Leimi commented on react/features/base/label/components/web/Label.tsx:
```diff
@@ -92,13 +92,27 @@ const Label = ({
+            onClick = { onClick }
+            onKeyPress = { onKeyPress }
+            role = { onClick ? 'button' : undefined }
+            tabIndex = { onClick ? 0 : undefined }>
```
I'd argue against it :)

It doesn't do the same thing. If the element is not interactive, we don't want any tabindex set to it. `-1` allows to programmatically focus the element. So, setting `-1` implies we want to be able to focus it in the code, while we don't want to here. If I see a `tabIndex={-1}` in some code, it tells me we do something with it somewhere. `undefined` shows us the attribute will not be generated and that's what we want.

Leimi commented on react/features/base/label/components/web/Label.tsx:
```diff
@@ -92,13 +92,27 @@ const Label = ({
-            onClick = { onClick }>
+            onClick = { onClick }
+            onKeyPress = { onKeyPress }
+            role = { onClick ? 'button' : undefined }
```
I'd also argue here that `undefined` would be better. The idea here is to say, "view this as a button if it's interactive, do as usual if not". `role="none"` removes the usual `role`. And I'd say it's better to have the usual role instead of `status` that can trigger automatic stuff depending on screen readers, something I think we wouldn't want in this low level component that could be used for different things.

Calinteodor commented on react/features/base/ui/components/web/ContextMenuItem.tsx:
```diff
@@ -148,7 +148,9 @@ const ContextMenuItem = ({
-            onKeyPress = { disabled ? undefined : onKeyPress }>
+            onKeyPress = { disabled ? undefined : onKeyPress }
+            role = 'button'
+            tabIndex = { disabled ? undefined : 0 }>
```
Got it. Thanks for the extra info @Leimi!
I converted this PR to draft, just in case you want to update it.

Leimi replied:
Well after rethinking about this I guess it's better to leave this as always a "button" role, the aria-disabled tells the user if it's actionable or not? I don't think it matters a lot to be honest.

Leimi commented:
## About the Dialog change

The code is now made so that it tries to focus back the element that triggered it. This works in some places, but in lots of places, it actually doesn't. Because lots of dialogs are opened via the overflow menu. I activate something in the overflow menu, it gets hidden, the dialog opens ; when I close it, it doesn't find back the hidden element so focus is set back to the body. One great thing to do would be open back the overflow menu if we know the dialog has been opened via keyboard, and focus back the button related to the dialog. Or, just focus the overflow menu button.

I think it would improve keyboard/screen reader navigation if we managed to do this. Some things are more important as for now but I'd love to try and go back on this to find a solution that doesn't necessarily demand huge code changes.

## About the Popover change

The code is now made so that when we open some popovers, they now trap the focus inside them, until something is activated inside them, or the escape key is used. After thinking about it, we should also mark the popovers has dialogs. So that screen reader users understand they are trapped in it and can use the esc key to go out. Because as for now it has some kind of dialog behavior without saying it to screen readers, which can be disturbing. Will make the change.

Calinteodor commented:
> ## About the Dialog change
> The code is now made so that it tries to focus back the element that triggered it. This works in some places, but in lots of places, it actually doesn't. Because lots of dialogs are opened via the overflow menu. I activate something in the overflow menu, it gets hidden, the dialog opens ; when I close it, it doesn't find back the hidden element so focus is set back to the body. One great thing to do would be open back the overflow menu if we know the dialog has been opened via keyboard, and focus back the button related to the dialog. Or, just focus the overflow menu button.
> 
> I think it would improve keyboard/screen reader navigation if we managed to do this. Some things are more important as for now but I'd love to try and go back on this to find a solution that doesn't necessarily demand huge code changes.
> 
> ## About the Popover change
> The code is now made so that when we open some popovers, they now trap the focus inside them, until something is activated inside them, or the escape key is used. After thinking about it, we should also mark the popovers as dialogs. So that screen reader users understand they are trapped in it and can use the esc key to go out. Because as for now it has some kind of dialog behavior without saying it to screen readers, which can be disturbing. Will make the change.

I checked out on your PR branch and tested the changes that you did and works as expected. Indeed the focus is trapped inside the dialogs and it exits only if we press esc key. Close button icons are not focused, but I think you took care of that in another PR that is in progress, right? Also, I think you are right about adding a role = 'dialog' to popovers, because they act like it. Thank you!

Leimi commented:
About the dialogs (not the popovers), I'd need to check again but no I didn't specifically handle a close-button related focus. Keyboard was already greatly handled in dialogs. I'd need to check again about this specifically.

It's mostly the fact that it didn't go back to where it was before opening the dialog that I changed :)

Leimi commented:
Hey, went back on this, updated to fix conflicts with updates on some parts I changed (audio/video settings, polls, dialog). I think we're good!

Leimi commented:
Hey, went back on this, updated to:

- fix conflicts with updates on some parts I changed (audio/video settings, polls, dialog)
- added https://github.com/jitsi/jitsi-meet/pull/12657/commits/77e24464a3c0991d2919e8154d4c8d98c943de88 as it is tied to a few changes I made
- updated Popover and Drawer components to add dialog-related html attributes. I added `headingId` and `headingLabel` props that match `aria-labelledby` and `aria-label` attributes on those dialogs so that you have the choice.

Calinteodor commented on react/features/base/popover/components/Popover.web.tsx:
```diff
@@ -34,6 +35,18 @@ interface IProps {
+     * The id of the dom element acting as the Popover label (matches aria-labelledby)
+     */
+    headingId?: string;
+
```
I think we can go with these props. You gave pretty explicit details.

Leimi commented:
Rebased and fixed the couple eslint issues :)

Calinteodor commented:
> Rebased and fixed the couple eslint issues :)

Seems like there are some issues inside main.json with some translations.

Leimi commented:
Oops :sweat_smile:  thought I only had to run `npm run lint` to check everything. Sorry!

Calinteodor commented:
> Oops 😅 thought I only had to run `npm run lint` to check everything. Sorry!

No worries!

Calinteodor commented:
> > Oops 😅 thought I only had to run `npm run lint` to check everything. Sorry!
> 
> No worries!

Still no luck :( Regarding the fact that your changes are web related: 
1. Delete tsconfig.native.json.
2. Run npm run lint and make dev to fix errors.
3. Re-add tsconfig.native.json.

Leimi commented:
Yeah I was gonna ask about those TSC errors because I don't understand why I have them… will try that thanks

Leimi commented:
I fixed the typescript errors :pray:

Leimi commented:
Is there anything more I should do to get this merged? :)

damencho commented:
jenkins test this please

Calinteodor commented:
> Is there anything more I should do to get this merged? :)
I will test again all your PRs in the morning. 
If the tests pass and everything works, we approve and merge!

Calinteodor commented:
This is good to merge! As you can see, from the checks section, no failing tests! 
Nice work, @Leimi!

P.S. I was able to test your PRs. Dunno what was that about, but VoiceOver works as expected and focus works.

Leimi commented:
Cooool :) thanks
